src/App.tsx

```diff
@@ -1,223 +1,298 @@
+  const [activeItems, setActiveItems] = useState<number>(0);
+  const [visibleLoader, setVisibleLoader] = useState(false);
+
+  let userId = 0;
+
+  if (user?.id) {
+    userId = user.id;
+  }
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
-    }
+    getTodos(userId)
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+        setActiveItems((prevItems) => {
+          const tempTodos = [...todosFromServer];
+          let tempItems = prevItems;
+
+          tempTodos.forEach(todo => {
+            if (!todo.completed) {
+              tempItems += 1;
+            }
+          });
+
+          setTodos(tempTodos);
+
+          return tempItems;
+        });
+      })
+      .catch(() => setErrorMessage('Unable to update todos'));
   }, []);
 
+  const findById = (todo: Todo) => {
+    const todoIndex = todos.findIndex(foundTodo => {
+      return foundTodo.id === todo.id;
+    });
+
+    return todoIndex;
+  };
+
+  const addNewTodo = (todo: Todo) => {
+    setTodos(prevTodos => [todo, ...prevTodos]);
+    setActiveItems(prevItems => prevItems + 1);
+  };
+
+  const todoDelete = (todoId: number) => {
+    deleteTodo(todoId)
+      .then(() => {
+        setVisibleLoader(false);
+      })
+      .then(() => {
+        setTodos(
+          todos.filter(userTodo => todoId !== userTodo.id),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      });
+    setActiveItems(prevItems => prevItems - 1);
+  };
+
+  const deleteCompletedTodos = () => {
+    const filteredTodos = todos.filter(todo => {
+      if (todo.completed) {
+        setVisibleLoader(true);
+
+        deleteTodo(todo.id)
+          .then(() => setVisibleLoader(false));
+      }
+
+      return todo.completed === false;
+    });
+
+    setTodos(filteredTodos);
+  };
+
+  const activeItemsCounter = (todo: Todo) => {
+    if (todo.completed) {
+      setActiveItems(prevItems => prevItems - 1);
+    } else {
+      setActiveItems(prevItems => prevItems + 1);
+    }
+  };
+
+  const updateCompleteTodo = (todo: Todo) => {
+    setVisibleLoader(true);
+
+    patchTodo(todo.id, { completed: !todo.completed })
+      .then(() => setVisibleLoader(false))
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => setVisibleLoader(false));
+
+    const newTodos = [...todos];
+
+    const todoIndex = findById(todo);
+
+    newTodos[todoIndex].completed = !todo.completed;
+
+    setTodos(newTodos);
+
+    activeItemsCounter(todo);
+  };
+
+  const updateAllCompleteTodos = () => {
+    const temp = [...todos];
+
+    temp.forEach(todo => {
+      setVisibleLoader(true);
+
+      patchTodo(todo.id, { completed: !todo.completed })
+        .then(() => {
+          setVisibleLoader(false);
+        })
+        .catch(() => setErrorMessage('Unable to update a todo'));
```

Comment:
it works uncorrectly
![image](https://user-images.githubusercontent.com/88807331/194544369-ae3374fb-eb14-4654-b518-f47d6a5b63dd.png)
![image](https://user-images.githubusercontent.com/88807331/194544391-b0d1f60a-b487-47d0-9e6d-cb6823e167aa.png)